lib/apphook.h

```diff
@@ -41,6 +41,8 @@ enum
 typedef void (*ApplicationHookFunc)(gint type, gpointer user_data);
 
 void register_application_hook(gint type, ApplicationHookFunc func, gpointer user_data);
+void register_future_application_hook(gint type, ApplicationHookFunc func, gpointer user_data);
+void append_application_hook(gint type, ApplicationHookFunc func, gpointer user_data);
 void app_startup(void);
```

Comment:
I dont see an implementation for the append function.

I dont understant the concept of "future", maybe a comment is due? Is there a better term? What does it do exactly?

Reply:
Well... the append_application_hook is removed, just didn't push every change.
Naming: maybe the best would be to rename the original register_application_hook as it hurts SRP: it not just register a callback but also run it sometimes.

Reply:
register_application_hook() would run that function only if that specific state has already passed. I can't remember the use-case exactly.

hmmm... This is the possible offender in afstreams.c:

```
      if (self->door_filename)
        {

          /* door creation is deferred, because it creates threads which is
           * not inherited through forks, and syslog-ng forks during
           * startup, but _after_ the configuration was initialized */

          register_application_hook(AH_POST_DAEMONIZED, afstreams_init_door, self);
        }

```

If syslog-ng is already in the background (e.g. during reload), that afstreams_init_door() should be called immediately. If it is starting up, it should be deferred.

Anyway, apphook probably deserves some care, I am not sure if the right time is now. But like I said, "future" hook is not a term  I easily understand.

Reply:
Ah, I see it was renamed to register_application_hook_without_checking_current_state()

It's better. Let me try to understand the logic again ...